File: src/JunoDetector.cc
#include "JunoDetector.hh"
#include "JunoPullTerms.hh"

#include <TMath.h>
#include <TFile.h>

JunoDetector::JunoDetector()
{
    m_Nproton = 1.43512e33;
    m_effciency = 0.82;

    // Calibration Paper 
    m_a = 0.0262;
    m_b = 0.0082;
    m_c = 0.0123;

}


JunoDetector::~JunoDetector()
{
    delete hNLnominal;
    delete hNL0;
    delete hNL1;
    delete hNL2;
    delete hNL3;
    delete hIBDtotXsec;
    delete fIBDdiffXsec;
    delete fEpositron;
}


void JunoDetector::LoadCommonInputs()
{
    TFile* ff = new TFile("JUNOInputs2022_01_06.root", "read");
    fIBDdiffXsec = (TF2*)ff->Get("dsigma_dcos_Enu_cos_DYB");
    fEpositron   = (TF2*)ff->Get("Epositron_Enu_cos_DYB");
    hNLnominal   = (TH1D*)ff->Get("positronScintNL");
    hNL0         = (TH1D*)ff->Get("positronScintNLpull0");
    hNL1         = (TH1D*)ff->Get("positronScintNLpull1");
    hNL2         = (TH1D*)ff->Get("positronScintNLpull2");
    hNL3         = (TH1D*)ff->Get("positronScintNLpull3");
    hIBDtotXsec  = (TH1D*)ff->Get("IBDXsec_VogelBeacom_DYB");
}


double JunoDetector::IBDdiffXsec(double Enu, double costheta)
{
    return fIBDdiffXsec->Eval(Enu, costheta);
}


double JunoDetector::PositronEnergy(double Enu, double costheta)
{
    return fEpositron->Eval(Enu, costheta);
}

double JunoDetector::IBDtotXsec(double Enu)
{
    return hIBDtotXsec->Interpolate(Enu);
}



double JunoDetector::Nonlinearity(double Edep)
{
    //return Edep * hNLnominal->Interpolate(Edep);
    return Edep * ( hNLnominal->Interpolate(Edep) + JunoPullTerms::alpha_l0*(hNL0->Interpolate(Edep) - hNLnominal->Interpolate(Edep)) + JunoPullTerms::alpha_l1*(hNL1->Interpolate(Edep) - hNLnominal->Interpolate(Edep)) + JunoPullTerms::alpha_l2*(hNL2->Interpolate(Edep) - hNLnominal->Interpolate(Edep)) + JunoPullTerms::alpha_l3*(hNL3->Interpolate(Edep) - hNLnominal->Interpolate(Edep)) );
}


double JunoDetector::Resolution(double Evis)
{
    double a = m_a * (1 + JunoPullTerms::alpha_ea);
    double b = m_b * (1 + JunoPullTerms::alpha_eb);
    double c = m_c * (1 + JunoPullTerms::alpha_ec);

    return TMath::Sqrt(a*a/Evis + b*b + c*c/Evis/Evis) * Evis;
}


double JunoDetector::DetectorResponse(double Enu, double costheta, double Ep)
{
    double Te    = PositronEnergy(Enu, costheta) - 0.511;
    double Edep  = Te + 1.022;
    double Evis  = Nonlinearity(Edep);
    double sigma = Resolution(Evis);
    
    return 1/(sigma * TMath::Sqrt(2*TMath::Pi())) * TMath::Exp(-(Ep - Evis) * (Ep - Evis) /2/sigma/sigma);
}

File: PYTHON/JunoDetector.py
import numpy as np
import ROOT



class JunoDetector(object):

    def __init__(self) -> None:


        # IBD differential cross section
        self.fIBDdiffXsec = ROOT.TF2()
        self.fEpositron   = ROOT.TF2()

        self.IBD_x = []
        self.IBD_y = []

        # detector configuration
        self.Nproton = 1.43512e33
        self.efficiency = 0.822
        self.hNL_x = []
        self.hNL_y = []
        self.a = 0.0261
        self.b = 0.0082
        self.c = 0.0123

    
    def GetNproton(self):
        return self.Nproton


    def GetEfficiency(self):
        return self.efficiency



    def CommonInput(self):
        ff = ROOT.TFile("JUNOInputs2022_01_06.root", "read")
        self.fIBDdiffXsec = ff.Get("dsigma_dcos_Enu_cos_DYB")
        self.fEpositron = ff.Get("Epositron_Enu_cos_DYB")
        hNL = ff.Get("positronScintNL")
        for i in range(hNL.GetNbinsX()):
            self.hNL_x.append(hNL.GetBinCenter(i+1))
            self.hNL_y.append(hNL.GetBinContent(i+1))
        hIBDXsec = ff.Get("IBDXsec_VogelBeacom_DYB")
        for i in range(hIBDXsec.GetNbinsX()):
            self.IBD_x.append(hIBDXsec.GetBinCenter(i+1))
            self.IBD_y.append(hIBDXsec.GetBinContent(i+1))


    
    def IBDdiffXsec(self, Enu, costheta):
        return self.fIBDdiffXsec.Eval(Enu, costheta)


    def IBDtotXsec(self, Enu):
        return np.interp(Enu, self.IBD_x, self.IBD_y)

    
    def nonlinearity(self, Ee):
        return np.interp(Ee+1.022, self.hNL_x, self.hNL_y) * Ee
    


    def resolution(self, Evis):
        return np.sqrt(self.a**2/Evis + self.b**2 + self.c**2/Evis**2) * Evis


    
    def detector_response(self, Enu, ct, Ep):
        Ee = self.fEpositron(Enu, ct) 
        Edep = Ee + 1.022
        Evis = self.nonlinearity(Edep)
        sigma = self.resolution(Evis)
        #print("neutrino energy", Enu)
        #print("costheta", ct)
        #print("positron KE", Ee)
        #print("positron Edep",  Edep)
        #print("positron Evis", Evis)
        #print("positron Esigma", sigma)
        #print("calc prompt E", Ep)

        return 1 / (np.sqrt(2*np.pi)*sigma) * np.exp(-(Ep - Evis)**2/(2*sigma**2)) 
    






